File: packages/sdk-typescript/src/evals/json.ts
import { createHash } from "node:crypto";
import type { JsonValue } from "../types.js";

export type JsonBounds = { bytes: number; nodes: number; depth: number };
export const valueBounds: JsonBounds = { bytes: 200_000, nodes: 20_000, depth: 32 };
export function aggregateBounds(bytes: number): JsonBounds {
  return { bytes, nodes: Math.ceil(bytes / 2), depth: valueBounds.depth + 8 };
}
const isText = (item: string) => item.isWellFormed() && !item.includes("\u0000");

/** A string's code points: its UTF-16 length less one for each surrogate pair. */
function codePoints(text: string): number {
  let count = text.length;
  for (let index = 0; index < text.length - 1; index++) {
    const unit = text.charCodeAt(index);
    if (unit < 0xd800 || unit > 0xdbff) continue;
    const next = text.charCodeAt(index + 1);
    if (next >= 0xdc00 && next <= 0xdfff) {
      count--;
      index++;
    }
  }
  return count;
}

/** Thrown within `json` when the byte bound is passed, which is checked last. */
class PastByteBound extends Error {}

/** Reject lossy JSON serialization before creating requests/checkpoints. */
export function json(value: unknown, requested: JsonBounds | number = valueBounds): JsonValue {
  const bounds = typeof requested === "number" ? { ...valueBounds, bytes: requested } : requested;
  const ancestors = new Set<object>();
  let nodes = 0;
  // The UTF-8 length of the JSON text, counted as the value is read, so a value too long to
  // serialize (past the runtime's longest string) is refused before `JSON.stringify` is asked to.
  let bytes = 0;
  // Once the byte bound is passed, the value is read again only to check it (below).
  let checking = false;
  // Text the second read has found valid is not checked again, so a string referenced many times
  // is scanned once: by content up to 16,383 UTF-16 units, and a longer string when it is the last
  // checked of its length. V8 hashes those by length alone, so a set of them would compare each
  // with every other of that length.
  const checkedText = new Set<string>();
  const checkedLong = new Map<number, string>();
  const text = (item: string) => {
    if (!checking) return isText(item);
    if (item.length <= 16_383 ? checkedText.has(item) : checkedLong.get(item.length) === item)
      return true;
    if (!isText(item)) return false;
    if (item.length <= 16_383) checkedText.add(item);
    else checkedLong.set(item.length, item);
    return true;
  };
  const charge = (amount: number) => {
    if (checking) return;
    bytes += amount;
    if (bytes > bounds.bytes) throw new PastByteBound();
  };
  // A string's JSON text is at least as long as the string, so one longer than what is left is
  // refused before it is escaped.
  const chargeText = (text: string) => {
    if (checking) return;
    if (text.length + 2 > bounds.bytes - bytes) throw new PastByteBound();
    charge(Buffer.byteLength(JSON.stringify(text)));
  };
  const visit = (item: unknown, depth: number): JsonValue => {
    if (++nodes > bounds.nodes || depth > bounds.depth)
      throw new RangeError("JSON exceeds depth/node limits");
    if (item === null || typeof item === "boolean") {
      charge(item === false ? 5 : 4);
      return item;
    }
    if (typeof item === "number" && Number.isFinite(item)) {
      charge(JSON.stringify(item).length);
      return item;
    }
    if (typeof item === "string" && text(item)) {
      chargeText(item);
      return item;
    }
    if (!item || typeof item !== "object" || ancestors.has(item))
      throw new TypeError("Expected finite JSON without cycles or invalid Unicode");
    ancestors.add(item);
    try {
      if (Array.isArray(item)) {
        // Each element is a value: an array longer than the values left is refused before its
        // keys are listed.
        if (item.length > bounds.nodes - nodes)
          throw new RangeError("JSON exceeds depth/node limits");
        if (Object.keys(item).length !== item.length)
          throw new TypeError("Sparse/extended arrays are not JSON");
        charge(item.length ? item.length + 1 : 2);
        return item.map((entry) => visit(entry, depth + 1));
      }
      if (Object.getPrototypeOf(item) !== Object.prototype && Object.getPrototypeOf(item) !== null)
        throw new TypeError("JSON objects must be plain objects");
      if (Object.getOwnPropertySymbols(item).length)
        throw new TypeError("JSON cannot contain symbol properties");
      const result: Record<string, JsonValue> = Object.create(null);
      // Each member is a value: an object with more than the values left is refused before its
      // keys are sorted.
      const keys = Object.keys(item);
      if (keys.length > bounds.nodes - nodes)
        throw new RangeError("JSON exceeds depth/node limits");
      if (!checking) {
        // Keys whose JSON text alone (at least their code points, two quotes and a colon each)
        // needs more bytes than are left pass the byte bound before they are sorted, as in the
        // Python SDK, so both read such an object the same way. A key has at least half its
        // UTF-16 length in code points, so a long one passes it without counting them.
        let left = bounds.bytes - bytes;
        for (const key of keys) {
          left -= 3;
          if (key.length / 2 > left) throw new PastByteBound();
          left -= codePoints(key);
          if (left < 0) throw new PastByteBound();
        }
        keys.sort();
      }
      charge(keys.length ? keys.length + 1 : 2);
      for (const key of keys) {
        if (!text(key))
          throw new TypeError("Expected finite JSON without cycles or invalid Unicode");
        chargeText(key);
        charge(1);
        const descriptor = Object.getOwnPropertyDescriptor(item, key)!;
        if (!("value" in descriptor)) throw new TypeError("JSON cannot contain accessors");
        result[key] = visit(descriptor.value, depth + 1);
      }
      return result;
    } finally {
      // Repeated siblings serialize independently; only an active ancestor is a cycle.
      ancestors.delete(item);
    }
  };
  try {
    const result = visit(value, 0);
    if (Buffer.byteLength(JSON.stringify(result)) > bounds.bytes)
      throw new RangeError("JSON exceeds byte limit");
    return result;
  } catch (error) {
    if (!(error instanceof PastByteBound)) throw error;
  }
  // The byte bound is checked last, as it was before the length was counted as the value is
  // read: a value past it that holds a value that is not JSON, or passes the value or depth bound,
  // is refused for that. So the value is read again with each object's keys in their own order,
  // and nothing is counted, escaped or sorted, nor text checked twice (`text`).
  checking = true;
  nodes = 0;
  visit(value, 0);
  throw new RangeError("JSON exceeds byte limit");
}
export function digest(value: unknown): string {
  return createHash("sha256")
    .update(JSON.stringify(json(value, aggregateBounds(8 * 1024 * 1024))))
    .digest("hex");
}
/** SHA-256 hex digest of scorer source, as declared in a `local_code` definition. */
export function sourceDigest(source: string | Uint8Array): string {
  return createHash("sha256").update(source).digest("hex");
}
export function uuid(value: string): string {
  if (
    typeof value !== "string" ||
    value.length !== 36 ||
    !/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(value)
  )
    throw new TypeError("Expected a UUID");
  return value;
}

File: packages/sdk-python/src/hue_sdk/evals/_json.py
from __future__ import annotations

import functools
import hashlib
import json
import math
import re
from typing import Any
from uuid import UUID


class _Missing:
    def __repr__(self) -> str:
        return "MISSING"


MISSING = _Missing()
VALUE_BYTES = 200_000
VALUE_NODES = 20_000
VALUE_DEPTH = 32


class JsonLimitError(ValueError):
    """JSON that is valid but beyond the size (``bytes``) or the count and depth
    (``structure``) bounds."""

    def __init__(self, limit: str, message: str) -> None:
        super().__init__(message)
        self.limit = limit


# A lone surrogate, found at C speed rather than character by character.
_SURROGATE = re.compile("[\ud800-\udfff]")


def _text(value: str) -> None:
    # An ASCII string, which CPython knows without scanning, cannot hold a surrogate.
    if "\0" in value or (not value.isascii() and _SURROGATE.search(value)):
        raise ValueError("JSON contains invalid Unicode or NUL.")


_ASTRAL = re.compile("[\U00010000-\U0010ffff]")
_ABOVE_SURROGATES = re.compile("[\ue000-\uffff]")


def _utf16_units(character: str) -> tuple[int, ...]:
    code = ord(character)
    if code < 0x10000:
        return (code,)
    code -= 0x10000
    return (0xD800 + (code >> 10), 0xDC00 + (code & 0x3FF))


def _compare_utf16(left: str, right: str) -> int:
    # Skip a shared prefix a chunk at a time, then find the first difference in the chunk that
    # differs by halves, all at C speed, so a long key is walked quickly and never copied whole.
    index, end = 0, min(len(left), len(right))
    while index < end and left[index : index + 65536] == right[index : index + 65536]:
        index += 65536
    size = 1 << min(16, max(end - index, 0).bit_length())
    while size > 1:
        size //= 2
        if left[index : index + size] == right[index : index + size]:
            index += size
    if index < end:
        return -1 if _utf16_units(left[index]) < _utf16_units(right[index]) else 1
    return len(left) - len(right)


def _utf16_order(keys: list[str]) -> list[str]:
    """Keys in UTF-16 code unit order, as JavaScript sorts them, without encoding them."""
    # Code point order differs only between a character above U+FFFF and one from U+E000.
    if not (
        any(_ASTRAL.search(key) for key in keys)
        and any(_ABOVE_SURROGATES.search(key) for key in keys)
    ):
        return sorted(keys)
    return sorted(keys, key=functools.cmp_to_key(_compare_utf16))


def _is_array_index(key: str) -> bool:
    """Whether JavaScript lists ``key`` among an object's array indices, ahead of its other keys."""
    return (
        len(key) <= 10
        and key.isascii()
        and key.isdigit()
        and (key == "0" or key[0] != "0")
        and int(key) < 2**32 - 1
    )


def _js_order(keys: list[str]) -> list[str]:
    """Keys in the order JavaScript lists an object's own keys: its array indices in numeric order,
    then the others as inserted."""
    indices = [key for key in keys if _is_array_index(key)]
    if not indices:
        return keys
    return sorted(indices, key=int) + [key for key in keys if not _is_array_index(key)]


class _PastByteBound(Exception):
    """Raised within ``json_value`` when the byte bound is passed, which is checked last."""


def json_value(value: Any, max_bytes: int = VALUE_BYTES) -> Any:
    """Validate before serialization; never coerce keys, NaN, dates or large Python integers.

    As the TypeScript SDK's ``json``, the value is read depth-first in order (an object's members
    by key), and each value is checked for its type before it is counted against the value, depth
    and byte bounds, so both SDKs refuse an output for the same reason. The byte count is that of
    the JSON text ``encode`` produces, kept as the value is read, so an output too large to
    serialize is refused without serializing it. The byte bound is checked last, as it was before
    it was counted as the value is read: past it, the value is read again, as in TypeScript, and
    one holding a value that is not JSON, or past the value or depth bound, is refused for that.
    """
    ancestors: set[int] = set()
    nodes = 0
    size = 0
    # Once the byte bound is passed, the value is read again only to check it (below).
    checking = False

    def charge(amount: int) -> None:
        nonlocal size
        if checking:
            return
        size += amount
        if size > max_bytes:
            raise _PastByteBound

    def charge_text(text: str) -> None:
        if checking:
            return
        # A string's JSON text is at least as long as the string, so one longer than what is left
        # is refused before it is escaped.
        if len(text) + 2 > max_bytes - size:
            raise _PastByteBound
        charge(len(json.dumps(text, ensure_ascii=False).encode("utf-8")))

    def visit(item: Any, depth: int) -> None:
        nonlocal nodes
        nodes += 1
        if nodes > VALUE_NODES or depth > VALUE_DEPTH:
            raise JsonLimitError("structure", "JSON exceeds depth or node limits.")
        if item is None or type(item) is bool:
            charge(5 if item is False else 4)
            return
        if type(item) is int and abs(item) <= 2**53 - 1:
            charge(len(str(item)))
            return
        if type(item) is float and math.isfinite(item):
            charge(len(json.dumps(item)))
            return
        if type(item) is str:
            _text(item)
            charge_text(item)
            return
        if type(item) not in (dict, list) or id(item) in ancestors:
            raise ValueError("Expected JSON without coercion or cycles.")
        # Each element or member is a value: a container with more than the values left is
        # refused before it is read.
        if len(item) > VALUE_NODES - nodes:
            raise JsonLimitError("structure", "JSON exceeds depth or node limits.")
        ancestors.add(id(item))
        try:
            if type(item) is list:
                charge(len(item) + 1 if item else 2)
                for child in item:
                    visit(child, depth + 1)
                return
            # Keys are checked but, as in the TypeScript SDK, not counted as values.
            if any(type(key) is not str for key in item):
                raise ValueError("JSON object keys must be strings.")
            # Keys whose JSON text alone (at least their code points, two quotes and a colon each)
            # needs more bytes than are left pass the byte bound before they are sorted, as in the
            # TypeScript SDK, so both read such an object the same way.
            key_bytes = 0
            for key in item if not checking else ():
                key_bytes += len(key) + 3
                if key_bytes > max_bytes - size:
                    raise _PastByteBound
            charge(len(item) + 1 if item else 2)
            for key in _js_order(list(item)) if checking else _utf16_order(list(item)):
                _text(key)
                charge_text(key)
                charge(1)
                visit(item[key], depth + 1)
        finally:
            # Repeated siblings serialize independently; only an active ancestor is a cycle.
            ancestors.discard(id(item))

    try:
        visit(value, 0)
        return value
    except _PastByteBound:
        pass
    # Read again with each object's keys in the order JavaScript lists them, as the TypeScript SDK
    # reads it again, and nothing is counted, escaped or sorted, which takes time linear in it.
    checking = True
    nodes = 0
    visit(value, 0)
    raise JsonLimitError("bytes", "JSON exceeds the byte limit.")


def encode(value: Any) -> bytes:
    return json.dumps(
        value, ensure_ascii=False, allow_nan=False, separators=(",", ":"), sort_keys=True
    ).encode("utf-8")


def digest(value: Any) -> str:
    return hashlib.sha256(encode(json_value(value, 8 * 1024 * 1024))).hexdigest()


def uuid(value: str) -> str:
    try:
        if not isinstance(value, str) or str(UUID(value)) != value.lower():
            raise ValueError
    except (ValueError, AttributeError):
        raise ValueError("Expected a UUID.") from None
    return value


def json_equal(left: Any, right: Any) -> bool:
    # Python bool is a subclass of int; JSON booleans and numbers remain distinct.
    if type(left) in (int, float) and type(right) in (int, float):
        return left == right
    if type(left) is not type(right):
        return False
    if isinstance(left, list):
        return len(left) == len(right) and all(
            json_equal(a, b) for a, b in zip(left, right, strict=True)
        )
    if isinstance(left, dict):
        return left.keys() == right.keys() and all(
            json_equal(left[key], right[key]) for key in left
        )
    return left == right
